### ISSUE TITLE
S3 Not responding to OPTIONS request when using presigned URL from the browser

### ISSUE THREAD

**Issue by aaronstillwell** [localstack/localstack#99]
I'm attempting to use a pre-signed URL from within the browser to upload to my localstack S3 service. Generating the URL itself is trivial, however when attempting to POST to the resulting url (e.g `http://localstack:4572/ps.local/test?X-Amz-Content-Sha256=UNSIGNED-PAYLOAD...`), the browser makes an OPTIONS preflight request, to which localstack does not respond.

It appears as if there is no way around this even with flexible CORS rules (see below). It's not therefore possible to directly upload files from the browser to localstack S3 using presigned URL's.
```
            $s3->putBucketCors([
                'Bucket' => 'myBucket',
                'CORSConfiguration' => [
                    'CORSRules' => [
                        [
                            'AllowedOrigins' => ['*'],
                            'AllowedMethods' => ['POST', 'GET', 'PUT'],
                            'MaxAgeSeconds' => 3000,
                            'AllowedHeaders' => ['*']
                        ]
                    ]
                ]
            ]);
```

If someone can advise on whether or not I've overlooked something I'll be happy to investigate further and follow up with a PR.

**Comment by whummer** [comment]
@aaronstillwell I think you are right, preflight `OPTIONS` requests are currently not implemented. If you could look into this and create a PR, that would be awesome. 

The S3 proxy listener function should be a good starting point:
https://github.com/atlassian/localstack/blob/master/localstack/mock/proxy/s3_listener.py#L200

We will also have to extend `generic_proxy.py` to forward `OPTIONS` requests to the listener - similar to the `do_HEAD` method for `HEAD` requests:
https://github.com/atlassian/localstack/blob/master/localstack/mock/generic_proxy.py#L53

Thanks!

**Comment by aaronstillwell** [comment]
https://bitbucket.org/atlassian/localstack/issues/20/s3-not-responding-to-options-request-when

**Comment by whummer** [comment]
Closing this issue as duplicate (tracked in the main Bitbucket repo)

**Comment by whummer** [comment]
CORS headers are now added by default, this has been fixed in commit 985a840f2229c09e7128e175048d47a5d9afd0b4 . Support for CORSRules is still missing
